design/design-doc-smaller-providers.md

```diff
@@ -0,0 +1,429 @@
+Kubernetes control plane and clients in the face of many CRDs. These
+improvements have been insufficient to alleviate community pain. The assumption
+that there won’t be “too many” CRDs is pervasive in the Kubernetes codebase.
+There’s no one easy fix, and the conservative speed at which cloud providers and
+their customers pick up new Kubernetes releases means it can take years for
+fixes to be deployed.
+
+Allowing folks to lower the ratio of installed-to-used CRDs has been proposed
+for some time. The Crossplane maintainers are now aligned that it’s necessary.
+We have explored the following options:
+
+* Make it possible to filter unnecessary CRDs by configuring a provider.
+* Break the biggest providers up into smaller, service-scoped providers.
+* Load CRDs into the API server lazily, on-demand.
+
+## Proposal
+
+I propose that provider maintainers break the largest providers into smaller,
+service-scoped providers. Each service-scoped provider would correspond to what
+is currently an API group within a provider. For example,
+`upbound/provider-aws-rds` would extend Crossplane with support for the 22
+managed resources (and thus 22 CRDs) in the `rds.aws.upbound.io` API group.
+
+### Overview
+
+I proposed we only break up the largest providers. For example:
+
+* `upbound/provider-aws` - Becomes ~150 smaller providers.
+* `upbound/provider-azure` - Becomes ~100 smaller providers.
+* `upbound/provider-gcp` - Becomes ~75 smaller providers.
+* `crossplane-contrib/provider-tencentcloud` - Becomes ~50 smaller providers.
+
+It's important to note that we don't expect anyone to be installing and
+operating all of these providers. Per [Community Impact](#community-impact), we
+expect folks to install ~10 providers on average.
+
+These providers would:
+
+* Be built from “monorepos” - the same repositories they are today, like
+  upbound/provider-aws.
+* Share a `ProviderConfig` by taking a dependency on a (e.g.)
+  `provider-aws-config` package.
+* Continue to use contemporary “compiled-in” cross-resource-references, short to
+  medium term.
+
+Provider generation tooling like [`upjet`][upjet] would be updated to support
+generating service-scoped providers.
+
+
+No changes would be required to Crossplane core - i.e. to the package manager.
+No changes would be required to the long tail of existing, smaller Crossplane
+providers such as `provider-terraform` and `provider-kubernetes`.
+
+The following are the high level pros and cons of the proposed approach.
+
+Pros
+
+* Fewer CRDs is the default user experience - no new concepts or configuration
+  required.
+* Requires no changes to Crossplane core - no waiting months for core changes to reach
+  beta and be on by default.
+* Provider work spread over many provider processes (i.e. pods) should scale
+  better.
+* Folks can create their own “meta-providers” (e.g. `acmeco/provider-aws`) that
+  pull in just the services they want. This is just packaging - no compiling
+  required.
```

Comment:
+1 In ACK land we are working on something similar [here](https://github.com/aws-controllers-k8s/community/issues/1360). Of course, we're coming to this from the opposite angle :) We started life with lots of individual service-specific controllers and got user feedback that installing lots of controllers was annoying!

Reply:
> got user feedback that installing lots of controllers was annoying

Could you link some of this feedback?

I'm wondering whether Crossplane's package manager might make it a little less onerous in our case.

Reply:
I went looking - here's some feedback https://github.com/aws-controllers-k8s/community/issues/1534.